cdc/kv/client.go

```diff
@@ -1013,27 +1022,40 @@ func (s *eventFeedSession) singleEventFeed(
 	var initialized uint32
 
 	matcher := newMatcher()
-
-	var ticker *time.Ticker
-	hangTime := time.Duration(0)
+	advanceCheckTicker := time.NewTicker(time.Second * 5)
+	lastReceivedEventTime := time.Now()
+	var lastResolvedTs uint64
 
 	for {
-		ticker = time.NewTicker(time.Minute)
 		var event *cdcpb.Event
 		var ok bool
 		select {
 		case <-ctx.Done():
 			return atomic.LoadUint64(&checkpointTs), ctx.Err()
-		case <-ticker.C:
-			hangTime += time.Minute
-			log.Warn("region not receiving event from tikv for too long time",
-				zap.Uint64("regionID", regionID), zap.Reflect("span", span), zap.Duration("duration", hangTime))
+		case <-advanceCheckTicker.C:
+			sinceLastEvent := time.Since(lastReceivedEventTime)
+			if sinceLastEvent > time.Second*20 {
+				log.Warn("region not receiving event from tikv for too long time",
+					zap.Uint64("regionID", regionID), zap.Reflect("span", span), zap.Duration("duration", sinceLastEvent))
+			}
+			version, err := s.kvStorage.CurrentVersion()
+			if err != nil {
+				return atomic.LoadUint64(&checkpointTs), errors.Trace(err)
+			}
+			currentTimeFromPD := oracle.GetTimeFromTS(version.Ver)
+			sinceLastResolvedTs := currentTimeFromPD.Sub(oracle.GetTimeFromTS(lastResolvedTs))
+			if sinceLastResolvedTs > time.Second*20 {
+				log.Warn("region not receiving resolved event from tikv for too long time, try to resolve lock",
+					zap.Uint64("regionID", regionID), zap.Reflect("span", span), zap.Duration("duration", sinceLastResolvedTs), zap.Uint64("lastResolvedTs", lastResolvedTs))
+				maxVersion := oracle.ComposeTS(oracle.GetPhysical(currentTimeFromPD.Add(-10*time.Second)), 0)
+				err = s.resolveLock(ctx, regionID, maxVersion)
+				if err != nil {
+					return atomic.LoadUint64(&checkpointTs), errors.Trace(err)
+				}
```

Comment:
I think it's better not to return error here (so do line 1043).  Instead, just continue the loop and retry resolveLock next time. In the current implementation if an error is returned from here, CDC will retry the region but TiKV may not know the region's eventFeed is canceled.